### PR TITLE
Add FAQ entry for hearbeat

### PR DESCRIPTION
**Pull request by himkt**
Although heartbeat mechanism is experimental, I think it would be better to add documentation.
Currently there is no documentation available for heartbeat usage and we have to read the release notes.

This explanation is largely derived from the past release notes.

🔗 https://github.com/optuna/optuna/releases/tag/v2.5.0
🔗 https://github.com/optuna/optuna/releases/tag/v2.6.0

### CODE REVIEW

File: docs/source/faq.rst [hunk 1]
```diff
@@ -494,3 +494,49 @@ There are two kinds of constrained optimizations, one with soft constraints and
...
+Can I monitor trials and make them failed automatically when they are killed by a scheduler?
+--------------------------------------------------------------------------------------------
+
+.. note::
+
+  Heartbeat mechanism is experimental. API would change in the future.
+
+Optuna supports monitoring trial heartbeats with RDB and Redis storages.
+Using heartbeat, if a process running a trial is killed by a scheduler in a cluster environment,
+Optuna will automatically change the state of the trial that was running on that process to `TrialState.FAIL`
```
Comment (nzw0301) [inline docs/source/faq.rst:508]:
```suggestion
Optuna will automatically change the state of the trial that was running on that process to :obj:`~optuna.trial.TrialState.FAIL`
```

File: docs/source/faq.rst [hunk 2]
```diff
@@ -494,3 +494,49 @@ There are two kinds of constrained optimizations, one with soft constraints and
...
+from `TrialState.RUNNING`.
```
Comment (nzw0301) [inline docs/source/faq.rst:509]:
```suggestion
from :obj:`~optuna.trial.TrialState.RUNNING`.
```

**Review by nzw0301 — COMMENTED** [review]
let me leave minor sphinx-ish comments.

**Comment by himkt** [comment]
Thank you, applied. :pray:

File: docs/source/faq.rst [hunk 3]
```diff
@@ -494,3 +494,49 @@ There are two kinds of constrained optimizations, one with soft constraints and
...
+
+    import optuna
+
+    def objective(trial):
+        (Very time-consuming computation)
+
+    # Recording heartbeats every 60 seconds.
+    # Other processes' trials where more than 120 seconds have passed
+    # since the last heartbeat was recorded will be automatically failed.
+    storage = optuna.storages.RDBStorage(url=..., heartbeat_interval=60, grace_period=120)
```
Comment (HideakiImamura) [inline docs/source/faq.rst:521]: For consistency below.
```suggestion
    storage = optuna.storages.RDBStorage(url="sqlite:///:memory:", heartbeat_interval=60, grace_period=120)
```

File: docs/source/faq.rst [hunk 4]
```diff
@@ -494,3 +494,49 @@ There are two kinds of constrained optimizations, one with soft constraints and
...
+    storage = optuna.storages.RDBStorage(
+        url="sqlite:///:memory:",
+        heartbeat_interval=60,
+        grace_period=120,
+        failed_trial_callback=RetryFailedTrialCallback(max_retry=3),
+    )
+
+    study = optuna.create_study(
+        storage=storage,
+    )
```
Comment (HideakiImamura) [inline docs/source/faq.rst:542]:
```suggestion
    study = optuna.create_study(storage=storage)
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I have small minor comments. PTAL.

File: docs/source/faq.rst [hunk 5]
```diff
@@ -494,3 +494,49 @@ There are two kinds of constrained optimizations, one with soft constraints and
 Soft constraints do not have to be satisfied, but an objective function is penalized if they are unsatisfied. On the other hand, hard constraints must be satisfied.
 
 Optuna is adopting the soft one and **DOES NOT** support the hard one. In other words, Optuna **DOES NOT** have built-in samplers for the hard constraints.
+
+
+Can I monitor trials and make them failed automatically when they are killed by a scheduler?
```
Comment (nzw0301) [inline docs/source/faq.rst:499]: Is the process killed by only a scheduler?

File: docs/source/faq.rst [hunk 6]
```diff
@@ -494,3 +494,49 @@ There are two kinds of constrained optimizations, one with soft constraints and
...
+--------------------------------------------------------------------------------------------
+
+.. note::
+
+  Heartbeat mechanism is experimental. API would change in the future.
+
+Optuna supports monitoring trial heartbeats with RDB and Redis storages.
```
Comment (nzw0301) [inline docs/source/faq.rst:506]: How about referring to [the Wikipedia page](https://en.wikipedia.org/wiki/Heartbeat_(computing)) or briefly explaining what is the heartbeat.

File: docs/source/faq.rst [hunk 7]
```diff
@@ -494,3 +494,49 @@ There are two kinds of constrained optimizations, one with soft constraints and
...
+        (Very time-consuming computation)
+
+    # Recording heartbeats every 60 seconds.
+    # Other processes' trials where more than 120 seconds have passed
+    # since the last heartbeat was recorded will be automatically failed.
+    storage = optuna.storages.RDBStorage(url=..., heartbeat_interval=60, grace_period=120)
+    study = optuna.create_study(storage=storage)
+    study.optimize(objective, n_trials=100)
+
+You can also execute a callback function to process the failed trial.
```
Comment (nzw0301) [inline docs/source/faq.rst:525]: Could you add when this callback tries to run the failed trial again?

**Review by nzw0301 — CHANGES_REQUESTED** [review]
I'm not familiar with the Heatbeat function, so I'd like to see a more informative explanation

File: docs/source/faq.rst (hunk 5 above)
Comment (himkt) [reply docs/source/faq.rst]: Nice point. I updated the title like 65d0121.

File: docs/source/faq.rst (hunk 6 above)
Comment (himkt) [reply docs/source/faq.rst]: da2454a adds the wikipedia link and improves the explanation. Please give me a feedback/

File: docs/source/faq.rst (hunk 7 above)
Comment (himkt) [reply docs/source/faq.rst:527]: Thank you. cf03ee3 adds it.

**Comment by himkt** [comment]
Sorry for the inactive @nzw0301 I didn't notice I received your kindly review. I revised based on the review.
No rush but please take a look.

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Review by nzw0301 — APPROVED** [review]
Brilliant. Thank you for your updates. LGTM.

**Pull request merged**